keps/sig-architecture/20190412-conformance-behaviors.md

```diff
@@ -0,0 +1,342 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This proposal modifies the conformance testing framework to be driven by a list
+of agreed upon behaviors. These behaviors are identified by processing of the
+API schemas, documentation, expert knowledge, and code examination. They are
+explicitly documented and tracked in the repository rather than in GitHub
+issues, allowing them to be reviewed and approved independently of the tests
+that evaluate them. Additionally it proposes new tooling to generate tests, test
+scaffolding, and test coverage reports.
+
+## Motivation
+
+It has proven difficult to measure how much of the expected Kubernetes behavior
+the current conformance tests cover. The current measurements are based upon
+identifying which API endpoints are exercised by the tests. The Kubernetes API
+is CRUD-oriented, and most of the client’s desired behavior is encapsulated in
+the payload of the create or update calls, not in the simple fact that those
+endpoints were hit. This means that even if a given endpoint is shown as
+covered, it’s impossible to know how much that tests the actual behavior.
+
+Coverage is measured this way because there is no single, explicit list of
+behaviors that comprise the expected behavior of a conformant cluster. These
+expectations are spread out across the existing design documents, KEPs, the user
+documentation, a subset of the e2e test, and the code itself. This makes it
+impossible to identify if the conformance suite provides a meaningful test of a
+cluster’s operation.
+
+Additionally, progress in writing and promoting tests has been slow and too much
+manual effort is involved. As a starting point, this proposal includes new
+tooling that uses the API schemas to identify expected behaviors and produce
+tests and test scaffolding to quickly cover those behaviors.
+
+### Goals
+
+* Enable separate review of behaviors and tests that evaluate those behaviors.
```

Comment:
+1 this is the main thing I wish to accomplish

- One group of humans (behavior-approvers) to ensure:
  - does this behavior describe the way kubernetes is supposed to operate
  - is this behavior something that is reasonably expected to work across all kubernetes clusters
- Another group of humans (test-approvers) to ensure:
  - does this test actually exercise this behavior
  - is this test written well according to our standards for conformance (in an ideal world, this would be lintable)

Today, promoting a test involves updating a file that is blocked on behavior-approvers.  Our only options to improve our velocity today are:
- train our test-approvers to reduce doublecheck review burden on behavior-approvers
- grow our pool of behavior-approvers

(I still think these approaches are healthy for the project, but we seem to have been unable to dedicate time or build sufficient trust here)

Reply:
To decouple things, promoting a test should not block on conformance behavior-approvers, but should block on conformance test-approvers.
- promotion should not involve updating behavior files
- promotion should involve update files under test-approver control

My best guess at the moment:
```
test/conformance
├── behaviors
│   ├── OWNERS # no-parent: true, approvers: behavior-approvers
│   └── {area}
│       ├── OWNERS # optional: reviewers: area-experts
│       └── {suite}.yaml
├── OWNERS # approvers: test-approvers
└── tests.yaml # promotion updates this file; tests MUST map to a behavior
```